webapp/src/main/java/com/box/l10n/mojito/service/tm/TMTextUnitStatisticService.java

```diff
@@ -0,0 +1,219 @@
+    /**
+     * logger
+     */
+    static Logger logger = LoggerFactory.getLogger(TMTextUnitStatisticService.class);
+
+    @Autowired
+    MeterRegistry meterRegistry;
+
+    @Autowired
+    TextUnitBatchMatcher textUnitBatchMatcher;
+
+    @Autowired
+    TextUnitDTOsCacheService textUnitDTOsCacheService;
+
+    @Autowired
+    TMTextUnitRepository tmTextUnitRepository;
+
+    @Autowired
+    TMTextUnitStatisticRepository tmTextUnitStatisticRepository;
+
+    @Autowired
+    TextUnitUtils textUnitUtils;
+
+    int batchSize = 5000;
+
+    /**
+     * Imports statistics information for a target text unit by matching it using name, content and comment information.
+     * Existing statistic data is overwritten by incoming data.
+     *
+     * @param locale                the locale of the target text unit
+     * @param asset                 the asset of the target text unit
+     * @param textUnitStatistics    the statistics of the text unit
+     * @return                      a PollableFuture
+     */
+    @Pollable(async = true, message = "Start importing text unit statistics.")
+    @Retryable(
+            value = SQLException.class,
+            maxAttempts = 2,
+            backoff = @Backoff(delay = 500, multiplier = 2))
+    public PollableFuture importStatistics(
+            Locale locale,
+            Asset asset,
+            List<ImportTextUnitStatisticsBody> textUnitStatistics) {
+        logger.debug("Import statistics for the text units for the given locale and asset.");
+
+        ImmutableMap<String, TextUnitDTO> textUnitDTOsByMD5 = textUnitDTOsCacheService.getTextUnitDTOsForAssetAndLocaleByMD5(
+                asset.getId(),
+                locale.getId(),
+                StatusFilter.ALL,
+                true,
+                UpdateType.ALWAYS);
+
+        Map<Pair<String, String>, List<TextUnitDTO>> textUnitDTOsByNameAndContent =
+                textUnitDTOsByMD5.values().stream()
+                        .collect(groupingBy(textUnitDTO -> new ImmutablePair<>(textUnitDTO.getName(), textUnitDTO.getSource())));
+
+        Map<ImportTextUnitStatisticsBody, ImmutableList<TextUnitDTO>> statisticToTextUnitDTOMap = textUnitStatistics.stream()
+                .filter(textUnitStatistic -> {
+                    ImmutableList<TextUnitDTO> matchedTextUnitDTOs = getMatchingTextUnitDTOs(textUnitStatistic, textUnitDTOsByMD5, textUnitDTOsByNameAndContent);
+                    if (matchedTextUnitDTOs == null) {
+                        logUnmatchedStatistics(asset, textUnitStatistic);
+                    }
+                    return matchedTextUnitDTOs != null;
+                })
+                .collect(Collectors.toMap(
+                        Function.identity(),
+                        textUnitStatistic -> getMatchingTextUnitDTOs(textUnitStatistic, textUnitDTOsByMD5, textUnitDTOsByNameAndContent),
+                        (textUnitDTOFirst, textUnitDTOLast) -> textUnitDTOLast));
```

Comment:
In this case the prioritization is actually undesirable - as want _all_ eligible matches.
The TextUnitBatchMatcher only allows mapping to one TextUnit.

The merge function selects the value for the key in the case of identical (duplicate) statistics - but the mapping function we're using is always deterministic and for duplicate keys it would always return the same list of matching DTOs back, thus we can return the equivalent value for the first key or the second - it doesn't really matter (they're the same) as long as the code doesn't trigger exceptions in that scenario.

Nit: I initially named those textUnitDTO**s**First (etc.) - not sure why I dropped the plurals - will add them back in for additional clarity.

Reply:
right, I read too fast, this dedup the statistic itself so all good. My comment around the `TextUnitBatchMatcher`  is more related to the `getMatchingTextUnitDTOs()` portion. I'm still wondering about the relevance of the matching by name and content.

Reply:
Changed the fallback matching to be more generic (by name) when content or comment are missing to accomodate all scenarios as discussed offline.